sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsublite/internal/SubscriberAssembler.java

```diff
@@ -18,47 +18,66 @@
 
 import static com.google.cloud.pubsublite.internal.ExtractStatus.toCanonical;
-import static com.google.cloud.pubsublite.internal.wire.ServiceClients.addDefaultMetadata;
 import static com.google.cloud.pubsublite.internal.wire.ServiceClients.addDefaultSettings;
+import static com.google.cloud.pubsublite.internal.wire.ServiceClients.getCallContext;
 import static java.util.concurrent.TimeUnit.MINUTES;
 
+import com.google.api.gax.rpc.ApiCallContext;
 import com.google.api.gax.rpc.ApiException;
+import com.google.cloud.pubsublite.AdminClient;
+import com.google.cloud.pubsublite.AdminClientSettings;
 import com.google.cloud.pubsublite.Offset;
 import com.google.cloud.pubsublite.Partition;
+import com.google.cloud.pubsublite.SubscriptionPath;
+import com.google.cloud.pubsublite.TopicPath;
 import com.google.cloud.pubsublite.internal.CursorClient;
 import com.google.cloud.pubsublite.internal.CursorClientSettings;
+import com.google.cloud.pubsublite.internal.ExtractStatus;
 import com.google.cloud.pubsublite.internal.wire.PubsubContext;
 import com.google.cloud.pubsublite.internal.wire.PubsubContext.Framework;
 import com.google.cloud.pubsublite.internal.wire.RoutingMetadata;
 import com.google.cloud.pubsublite.internal.wire.SubscriberBuilder;
 import com.google.cloud.pubsublite.internal.wire.SubscriberFactory;
-import com.google.cloud.pubsublite.proto.CommitCursorRequest;
 import com.google.cloud.pubsublite.proto.Cursor;
 import com.google.cloud.pubsublite.proto.SeekRequest;
-import com.google.cloud.pubsublite.v1.CursorServiceClient;
-import com.google.cloud.pubsublite.v1.CursorServiceSettings;
 import com.google.cloud.pubsublite.v1.SubscriberServiceClient;
 import com.google.cloud.pubsublite.v1.SubscriberServiceSettings;
+import java.util.concurrent.ConcurrentHashMap;
 import org.apache.beam.sdk.io.gcp.pubsublite.SubscriberOptions;
 
 class SubscriberAssembler {
   private static final Framework FRAMEWORK = Framework.of("BEAM");
+  private static final ConcurrentHashMap<SubscriptionPath, TopicPath> KNOWN_PATHS =
+      new ConcurrentHashMap<>();
+  private static final ConcurrentHashMap<SubscriptionPath, SubscriberServiceClient> SUB_CLIENTS =
+      new ConcurrentHashMap<>();
+  private static final ConcurrentHashMap<SubscriptionPath, CursorClient> CURSOR_CLIENTS =
+      new ConcurrentHashMap<>();
+
   private final SubscriberOptions options;
   private final Partition partition;
 
-  SubscriberAssembler(SubscriberOptions options, Partition partition) {
-    this.options = options;
-    this.partition = partition;
+  private static TopicPath lookupTopicPath(SubscriptionPath subscriptionPath) {
+    try (AdminClient adminClient =
+        AdminClient.create(
+            AdminClientSettings.newBuilder()
+                .setRegion(subscriptionPath.location().extractRegion())
+                .build())) {
+      return TopicPath.parse(
+          adminClient.getSubscription(subscriptionPath).get(1, MINUTES).getTopic());
+    } catch (Throwable t) {
```

Comment:
```suggestion
    } catch (Exception e) {
```

Don't attempt to catch Errors unless there is a special reason to do so. (This pattern is already widely used so it's fine to bulk edit in a separate PR.)

https://docs.oracle.com/javase/8/docs/api/java/lang/Error.html